FTBFS: variable length arrays are treated as errors

### Operating System Info

Other

### Other OS

Debian

### OBS Studio Version

29.1.3

### OBS Studio Version (Other)

_No response_

### OBS Studio Log URL

n.n.

### OBS Studio Crash Log URL

_No response_

### Expected Behavior

i expect to be able to compile obs-studio with all it's bundled plugins

### Current Behavior

build fails with:

```
cd /build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/obj-x86_64-linux-gnu/plugins/linux-pipewire && /usr/bin/ccache /usr/bin/cc -DENABLE_HEVC -DHAVE_OBSCONFIG_H -Dlinux_pipewire_EXPORTS -I/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/libobs -I/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/obj-x86_64-linux-gnu/config -I/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/deps/glad/include -isystem /usr/include/pipewire-0.3 -isystem /usr/include/spa-0.2 -isystem /usr/include/glib-2.0 -isystem /usr/lib/x86_64-linux-gnu/glib-2.0/include -isystem /usr/include/libmount -isystem /usr/include/blkid -isystem /usr/include/gio-unix-2.0 -g -O2 -ffile-prefix-map=/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg=. -fstack-protector-strong -Wformat -Werror=format-security -DSIMDE_ENABLE_OPENMP -fopenmp-simd -O3 -Wdate-time -D_FORTIFY_SOURCE=2 -DFFMPEG_MUX_FIXED=\"/usr/lib/x86_64-linux-gnu/obs-plugins/obs-ffmpeg/obs-ffmpeg-mux\" -std=gnu17 -fPIC -Werror -Wextra -Wvla -Wswitch -Wno-error=switch -Wformat -Wformat-security -Wunused-parameter -Wno-unused-function -Wno-missing-field-initializers -fno-strict-aliasing -Werror-implicit-function-declaration -Wno-missing-braces -Wno-error=maybe-uninitialized -mmmx -msse -msse2 -I/usr/include/pipewire-0.3 -I/usr/include/spa-0.2 -D_REENTRANT -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/gio-unix-2.0 -pthread -MD -MT plugins/linux-pipewire/CMakeFiles/linux-pipewire.dir/pipewire.c.o -MF CMakeFiles/linux-pipewire.dir/pipewire.c.o.d -o CMakeFiles/linux-pipewire.dir/pipewire.c.o -c /build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/plugins/linux-pipewire/pipewire.c
/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/plugins/linux-pipewire/pipewire.c: In function ‘on_process_cb’:
/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/plugins/linux-pipewire/pipewire.c:596:17: error: ISO C90 forbids variable length array ‘offsets’ [-Werror=vla]
  596 |                 uint32_t offsets[planes];
      |                 ^~~~~~~~
/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/plugins/linux-pipewire/pipewire.c:597:17: error: ISO C90 forbids variable length array ‘strides’ [-Werror=vla]
  597 |                 uint32_t strides[planes];
      |                 ^~~~~~~~
/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/plugins/linux-pipewire/pipewire.c:598:17: error: ISO C90 forbids variable length array ‘modifiers’ [-Werror=vla]
  598 |                 uint64_t modifiers[planes];
      |                 ^~~~~~~~
/build/obs-studio-2j3wxb/obs-studio-29.1.3+dfsg/plugins/linux-pipewire/pipewire.c:599:17: error: ISO C90 forbids variable length array ‘fds’ [-Werror=vla]
  599 |                 int fds[planes];
      |                 ^~~
cc1: all warnings being treated as errors
```

The error is obviously triggered because
- the code uses variable length arrays https://github.com/obsproject/obs-studio/blob/19d7a96de38e40d89cc953db84452b53ec63487d/plugins/linux-pipewire/pipewire.c#L614-L617
- the code is compiled with `-Werror -Wvla`
presumable because of https://github.com/obsproject/obs-studio/blob/19d7a96de38e40d89cc953db84452b53ec63487d/cmake/Modules/CompilerConfig.cmake#L110-L113

cmake is invoked with:
```sh
cd obj-x86_64-linux-gnu && cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DFETCHCONTENT_FULLY_DISCONNECTED=ON -DCMAKE_INSTALL_RUNSTATEDIR=/run -DCMAKE_SKIP_INSTALL_ALL_DEPENDENCY=ON "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DOBS_MULTIARCH_SUFFIX=/x86_64-linux-gnu -DUNIX_STRUCTURE=TRUE -DCMAKE_SKIP_RPATH=TRUE -DDISABLE_UPDATE_MODULE=TRUE -DBUILD_VST=OFF -DENABLE_AJA=OFF -DENABLE_SNDIO=OFF -DENABLE_JACK=TRUE -DOBS_VERSION_OVERRIDE=29.1.3.1-1 -DAPPDATA_RELEASE_DATE=2023-06-27 ..
```

### Steps to Reproduce

https://salsa.debian.org/multimedia-team/obs-studio/-/jobs/4368844

### Anything else we should know?

given that these compiler warnings and errors are mostly for development of obs-studio, should/could they be disabled for release-builds?

--- a/libobs/obs-cocoa.m
+++ b/libobs/obs-cocoa.m
@@ -462,7 +462,7 @@ static bool code_to_str(int code, struct dstr *str)
 void obs_key_to_str(obs_key_t key, struct dstr *str)
 {
     const UniCharCount max_length = 16;
-    UniChar buffer[max_length];
+    UniChar buffer[16];
 
     if (localized_key_to_str(key, str))
         return;

--- a/libobs/util/sse-intrin.h
+++ b/libobs/util/sse-intrin.h
@@ -17,6 +17,8 @@
 
 #pragma once
 
+#include "c99defs.h"
+
 #if (defined(_MSC_VER) || defined(__MINGW32__)) && \
 	((defined(_M_X64) && !defined(_M_ARM64EC)) || defined(_M_IX86))
 #include <emmintrin.h>
@@ -25,5 +27,7 @@
 #define WIN32_LEAN_AND_MEAN
 #endif
 #define SIMDE_ENABLE_NATIVE_ALIASES
+PRAGMA_WARN_PUSH
 #include "simde/x86/sse2.h"
+PRAGMA_WARN_POP
 #endif

--- a/plugins/linux-pipewire/pipewire.c
+++ b/plugins/linux-pipewire/pipewire.c
@@ -811,12 +811,21 @@ static void process_video_sync(obs_pipewire_stream *obs_pw_stream)
 
 	if (buffer->datas[0].type == SPA_DATA_DmaBuf) {
 		uint32_t planes = buffer->n_datas;
-		uint32_t offsets[planes];
-		uint32_t strides[planes];
-		uint64_t modifiers[planes];
-		int fds[planes];
+		DARRAY(uint32_t) offsets;
+		DARRAY(uint32_t) strides;
+		DARRAY(uint64_t) modifiers;
+		DARRAY(int) fds;
 		bool use_modifiers;
 		bool corrupt = false;
+
+		da_init(offsets);
+		da_reserve(offsets, planes);
+		da_init(strides);
+		da_reserve(strides, planes);
+		da_init(modifiers);
+		da_reserve(modifiers, planes);
+		da_init(fds);
+		da_reserve(fds, planes);
 
 #ifdef DEBUG_PIPEWIRE
 		blog(LOG_DEBUG,
@@ -838,11 +847,13 @@ static void process_video_sync(obs_pipewire_stream *obs_pw_stream)
 		}
 
 		for (uint32_t plane = 0; plane < planes; plane++) {
-			fds[plane] = buffer->datas[plane].fd;
-			offsets[plane] = buffer->datas[plane].chunk->offset;
-			strides[plane] = buffer->datas[plane].chunk->stride;
-			modifiers[plane] =
-				obs_pw_stream->format.info.raw.modifier;
+			da_push_back(fds, &buffer->datas[plane].fd);
+			da_push_back(offsets,
+				     &buffer->datas[plane].chunk->offset);
+			da_push_back(strides,
+				     &buffer->datas[plane].chunk->stride);
+			da_push_back(modifiers,
+				     &obs_pw_stream->format.info.raw.modifier);
 			corrupt |= (buffer->datas[plane].chunk->flags &
 				    SPA_CHUNK_FLAG_CORRUPTED) > 0;
 		}
@@ -850,6 +861,12 @@ static void process_video_sync(obs_pipewire_stream *obs_pw_stream)
 		if (corrupt) {
 			blog(LOG_DEBUG,
 			     "[pipewire] buffer contains corrupted data");
+
+			da_free(offsets);
+			da_free(strides);
+			da_free(modifiers);
+			da_free(fds);
+
 			goto read_metadata;
 		}
 
@@ -860,8 +877,14 @@ static void process_video_sync(obs_pipewire_stream *obs_pw_stream)
 		obs_pw_stream->texture = gs_texture_create_from_dmabuf(
 			obs_pw_stream->format.info.raw.size.width,
 			obs_pw_stream->format.info.raw.size.height,
-			format_data.drm_format, GS_BGRX, planes, fds, strides,
-			offsets, use_modifiers ? modifiers : NULL);
+			format_data.drm_format, GS_BGRX, planes, fds.array,
+			strides.array, offsets.array,
+			use_modifiers ? modifiers.array : NULL);
+
+		da_free(offsets);
+		da_free(strides);
+		da_free(modifiers);
+		da_free(fds);
 
 		if (obs_pw_stream->texture == NULL) {
 			remove_modifier_from_format(

--- a/plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
+++ b/plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
@@ -369,27 +369,25 @@ static bool virtualcam_output_start(void *data)
                                            .mScope = kCMIOObjectPropertyScopeGlobal,
                                            .mElement = kCMIOObjectPropertyElementMain};
         CMIOObjectGetPropertyDataSize(kCMIOObjectSystemObject, &address, 0, NULL, &size);
-        size_t num_devices = size / sizeof(CMIOObjectID);
-        CMIOObjectID cmio_devices[num_devices];
-        CMIOObjectGetPropertyData(kCMIOObjectSystemObject, &address, 0, NULL, size, &used, &cmio_devices);
+        NSMutableData *cmioDevices = [NSMutableData dataWithLength:size];
+        void *device_data = [cmioDevices mutableBytes];
+        CMIOObjectGetPropertyData(kCMIOObjectSystemObject, &address, 0, NULL, size, &used, device_data);
 
         vcam->deviceID = 0;
-
         NSString *OBSVirtualCamUUID = [[NSBundle bundleWithIdentifier:@"com.obsproject.mac-virtualcam"]
             objectForInfoDictionaryKey:@"OBSCameraDeviceUUID"];
-        for (size_t i = 0; i < num_devices; i++) {
-            CMIOObjectID cmio_device = cmio_devices[i];
+        for (size_t i = 0; i < (used * sizeof(CMIOObjectID)); i++) {
+            CMIOObjectID cmioDevice;
+            [cmioDevices getBytes:&cmioDevice range:NSMakeRange(i * sizeof(CMIOObjectID), sizeof(CMIOObjectID))];
+
             address.mSelector = kCMIODevicePropertyDeviceUID;
-
             UInt32 device_name_size;
-            CMIOObjectGetPropertyDataSize(cmio_device, &address, 0, NULL, &device_name_size);
-
+            CMIOObjectGetPropertyDataSize(cmioDevice, &address, 0, NULL, &device_name_size);
             CFStringRef uid;
-            CMIOObjectGetPropertyData(cmio_device, &address, 0, NULL, device_name_size, &used, &uid);
-
+            CMIOObjectGetPropertyData(cmioDevice, &address, 0, NULL, device_name_size, &used, &uid);
             const char *uid_string = CFStringGetCStringPtr(uid, kCFStringEncodingUTF8);
             if (uid_string && strcmp(uid_string, OBSVirtualCamUUID.UTF8String) == 0) {
-                vcam->deviceID = cmio_device;
+                vcam->deviceID = cmioDevice;
                 CFRelease(uid);
                 break;
             } else {
@@ -404,11 +402,15 @@ static bool virtualcam_output_start(void *data)
 
         address.mSelector = kCMIODevicePropertyStreams;
         CMIOObjectGetPropertyDataSize(vcam->deviceID, &address, 0, NULL, &size);
-        CMIOStreamID stream_ids[(size / sizeof(CMIOStreamID))];
+        NSMutableData *streamIds = [NSMutableData dataWithLength:size];
+        void *stream_data = [streamIds mutableBytes];
+        CMIOObjectGetPropertyData(vcam->deviceID, &address, 0, NULL, size, &used, stream_data);
 
-        CMIOObjectGetPropertyData(vcam->deviceID, &address, 0, NULL, size, &used, &stream_ids);
-
-        vcam->streamID = stream_ids[1];
+        if (used < (2 * sizeof(CMIOStreamID))) {
+            obs_output_set_last_error(vcam->output, obs_module_text("Error.SystemExtension.CameraNotStarted"));
+            return false;
+        }
+        [streamIds getBytes:&vcam->streamID range:NSMakeRange(sizeof(CMIOStreamID), sizeof(CMIOStreamID))];
 
         CMIOStreamCopyBufferQueue(
             vcam->streamID, [](CMIOStreamID, void *, void *) {

